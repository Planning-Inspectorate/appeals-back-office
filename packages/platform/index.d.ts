import { HandlerFunction } from 'got';

declare module '@pins/platform' {
	export function createTtlHandler(): HandlerFunction;
	export function validatePostcode(value: string): boolean;
	export function validatePastDate(value: Date | string | number): boolean;
	export function validateFutureDate(value: Date | string | number): boolean;
	export function yesterday(): Date;
}

export interface LocalPlanningDepartment {
	FID: number;
	LPA21CD: string;
	LPA21NM: string;
	Co_terminous: 'Y' | 'N';
}
