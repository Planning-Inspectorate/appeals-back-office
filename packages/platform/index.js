export * from './http/ttl.js';
export * from './util/date.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
