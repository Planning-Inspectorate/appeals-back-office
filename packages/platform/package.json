{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=22.0.0 <= 22.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci",
    "test:cov": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci --coverage",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.3",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.4",
    "@types/lodash-es": "^4.17.12",
    "cross-env": "^7.0.3",
    "jest": "^29.7.0",
    "type-fest": "^4.37.0"
  }
}
