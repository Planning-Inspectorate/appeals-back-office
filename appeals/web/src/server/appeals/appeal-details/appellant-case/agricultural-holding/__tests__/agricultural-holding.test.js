// @ts-nocheck
import { parseHtml } from '@pins/platform';
import supertest from 'supertest';
import { appealData, appellantCaseDataNotValidated } from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';
import nock from 'nock';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details';

describe('agricultural-holding', () => {
	beforeEach(installMockApi);
	afterEach(teardown);

	describe('GET /change', () => {
		it('should render the change part of agricultural holding page with "no" radio option checked if agriculturalHolding.isPartOfAgriculturalHolding is false', async () => {
			nock('http://test/')
				.get(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
				.reply(200, {
					...appellantCaseDataNotValidated,
					agriculturalHolding: {
						...appellantCaseDataNotValidated.agriculturalHolding,
						isPartOfAgriculturalHolding: false
					}
				});

			const response = await request.get(`${baseUrl}/1/appellant-case/agricultural-holding/change`);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain('Change part of agricultural holding</h1>');
			expect(unprettifiedElement.innerHTML).toContain(
				'name="partOfAgriculturalHoldingRadio" type="radio" value="no" checked=null>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="partOfAgriculturalHoldingRadio" type="radio" value="yes">'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});

		it('should render the change part of agricultural holding page with "yes" radio option checked if agriculturalHolding.isPartOfAgriculturalHolding is true', async () => {
			nock('http://test/')
				.get(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
				.reply(200, {
					...appellantCaseDataNotValidated,
					agriculturalHolding: {
						...appellantCaseDataNotValidated.agriculturalHolding,
						isPartOfAgriculturalHolding: true
					}
				});

			const response = await request.get(`${baseUrl}/1/appellant-case/agricultural-holding/change`);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain('Change part of agricultural holding</h1>');
			expect(unprettifiedElement.innerHTML).toContain(
				'name="partOfAgriculturalHoldingRadio" type="radio" value="no">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="partOfAgriculturalHoldingRadio" type="radio" value="yes" checked=null>'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});
	});

	describe('POST /change', () => {
		const validValues = ['Yes', 'No'];

		for (const validValue of validValues) {
			it(`should call appellant cases PATCH endpoint and redirect to the appellant case page if "${validValue}" is selected`, async () => {
				const mockAppellantCasesPatchEndpoint = nock('http://test/')
					.patch(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
					.reply(200, {});

				const response = await request
					.post(`${baseUrl}/1/appellant-case/agricultural-holding/change`)
					.send({
						partOfAgriculturalHoldingRadio: validValue.toLowerCase()
					});

				expect(mockAppellantCasesPatchEndpoint.isDone()).toBe(true);
				expect(response.statusCode).toBe(302);
				expect(response.text).toBe(
					'Found. Redirecting to /appeals-service/appeal-details/1/appellant-case'
				);
			});
		}
	});

	describe('GET /tenant/change', () => {
		it('should render the change tenant of agricultural holding page with "no" radio option checked if agriculturalHolding.isTenant is false', async () => {
			nock('http://test/')
				.get(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
				.reply(200, {
					...appellantCaseDataNotValidated,
					agriculturalHolding: {
						...appellantCaseDataNotValidated.agriculturalHolding,
						isTenant: false
					}
				});

			const response = await request.get(
				`${baseUrl}/1/appellant-case/agricultural-holding/tenant/change`
			);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain('Change tenant of agricultural holding</h1>');
			expect(unprettifiedElement.innerHTML).toContain(
				'name="tenantOfAgriculturalHoldingRadio" type="radio" value="no" checked=null>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="tenantOfAgriculturalHoldingRadio" type="radio" value="yes">'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});

		it('should render the change tenant of agricultural holding page with "yes" radio option checked if agriculturalHolding.isTenant is true', async () => {
			nock('http://test/')
				.get(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
				.reply(200, {
					...appellantCaseDataNotValidated,
					agriculturalHolding: {
						...appellantCaseDataNotValidated.agriculturalHolding,
						isTenant: true
					}
				});

			const response = await request.get(
				`${baseUrl}/1/appellant-case/agricultural-holding/tenant/change`
			);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain('Change tenant of agricultural holding</h1>');
			expect(unprettifiedElement.innerHTML).toContain(
				'name="tenantOfAgriculturalHoldingRadio" type="radio" value="no">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="tenantOfAgriculturalHoldingRadio" type="radio" value="yes" checked=null>'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});
	});

	describe('POST /tenant/change', () => {
		const validValues = ['Yes', 'No'];

		for (const validValue of validValues) {
			it(`should call appellant cases PATCH endpoint and redirect to the appellant case page if "${validValue}" is selected`, async () => {
				const mockAppellantCasesPatchEndpoint = nock('http://test/')
					.patch(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
					.reply(200, {});

				const response = await request
					.post(`${baseUrl}/1/appellant-case/agricultural-holding/tenant/change`)
					.send({
						tenantOfAgriculturalHoldingRadio: validValue.toLowerCase()
					});

				expect(mockAppellantCasesPatchEndpoint.isDone()).toBe(true);
				expect(response.statusCode).toBe(302);
				expect(response.text).toBe(
					'Found. Redirecting to /appeals-service/appeal-details/1/appellant-case'
				);
			});
		}
	});

	describe('GET /other-tenants/change', () => {
		it('should render the change other tenants of agricultural holding page with "no" radio option checked if agriculturalHolding.hasOtherTenants is false', async () => {
			nock('http://test/')
				.get(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
				.reply(200, {
					...appellantCaseDataNotValidated,
					agriculturalHolding: {
						...appellantCaseDataNotValidated.agriculturalHolding,
						hasOtherTenants: false
					}
				});

			const response = await request.get(
				`${baseUrl}/1/appellant-case/agricultural-holding/other-tenants/change`
			);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain(
				'Change other tenants of agricultural holding</h1>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="otherTenantsOfAgriculturalHoldingRadio" type="radio" value="no" checked=null>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="otherTenantsOfAgriculturalHoldingRadio" type="radio" value="yes">'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});

		it('should render the change other tenants of agricultural holding page with "yes" radio option checked if agriculturalHolding.hasOtherTenants is true', async () => {
			nock('http://test/')
				.get(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
				.reply(200, {
					...appellantCaseDataNotValidated,
					agriculturalHolding: {
						...appellantCaseDataNotValidated.agriculturalHolding,
						hasOtherTenants: true
					}
				});

			const response = await request.get(
				`${baseUrl}/1/appellant-case/agricultural-holding/other-tenants/change`
			);
			const element = parseHtml(response.text);

			expect(element.innerHTML).toMatchSnapshot();

			const unprettifiedElement = parseHtml(response.text, { skipPrettyPrint: true });

			expect(unprettifiedElement.innerHTML).toContain(
				'Change other tenants of agricultural holding</h1>'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="otherTenantsOfAgriculturalHoldingRadio" type="radio" value="no">'
			);
			expect(unprettifiedElement.innerHTML).toContain(
				'name="otherTenantsOfAgriculturalHoldingRadio" type="radio" value="yes" checked=null>'
			);
			expect(unprettifiedElement.innerHTML).toContain('Continue</button>');
		});
	});

	describe('POST /other-tenants/change', () => {
		const validValues = ['Yes', 'No'];

		for (const validValue of validValues) {
			it(`should call appellant cases PATCH endpoint and redirect to the appellant case page if "${validValue}" is selected`, async () => {
				const mockAppellantCasesPatchEndpoint = nock('http://test/')
					.patch(`/appeals/1/appellant-cases/${appealData.appellantCaseId}`)
					.reply(200, {});

				const response = await request
					.post(`${baseUrl}/1/appellant-case/agricultural-holding/other-tenants/change`)
					.send({
						otherTenantsOfAgriculturalHoldingRadio: validValue.toLowerCase()
					});

				expect(mockAppellantCasesPatchEndpoint.isDone()).toBe(true);
				expect(response.statusCode).toBe(302);
				expect(response.text).toBe(
					'Found. Redirecting to /appeals-service/appeal-details/1/appellant-case'
				);
			});
		}
	});
});
