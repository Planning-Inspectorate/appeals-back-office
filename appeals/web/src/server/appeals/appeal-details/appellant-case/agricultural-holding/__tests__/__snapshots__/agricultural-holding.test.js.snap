// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agricultural-holding GET /change should render the change part of agricultural holding page with "no" radio option checked if agriculturalHolding.isPartOfAgriculturalHolding is false 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change part of agricultural holding</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="partOfAgriculturalHoldingRadio"
                            name="partOfAgriculturalHoldingRadio" type="radio" value="yes">
                            <label class="govuk-label govuk-radios__label" for="partOfAgriculturalHoldingRadio">Yes</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="partOfAgriculturalHoldingRadio-2"
                            name="partOfAgriculturalHoldingRadio" type="radio" value="no" checked>
                            <label class="govuk-label govuk-radios__label" for="partOfAgriculturalHoldingRadio-2">No</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`agricultural-holding GET /change should render the change part of agricultural holding page with "yes" radio option checked if agriculturalHolding.isPartOfAgriculturalHolding is true 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change part of agricultural holding</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="partOfAgriculturalHoldingRadio"
                            name="partOfAgriculturalHoldingRadio" type="radio" value="yes" checked>
                            <label class="govuk-label govuk-radios__label" for="partOfAgriculturalHoldingRadio">Yes</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="partOfAgriculturalHoldingRadio-2"
                            name="partOfAgriculturalHoldingRadio" type="radio" value="no">
                            <label class="govuk-label govuk-radios__label" for="partOfAgriculturalHoldingRadio-2">No</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`agricultural-holding GET /other-tenants/change should render the change other tenants of agricultural holding page with "no" radio option checked if agriculturalHolding.hasOtherTenants is false 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change other tenants of agricultural holding</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="otherTenantsOfAgriculturalHoldingRadio"
                            name="otherTenantsOfAgriculturalHoldingRadio" type="radio" value="yes">
                            <label class="govuk-label govuk-radios__label" for="otherTenantsOfAgriculturalHoldingRadio">Yes</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="otherTenantsOfAgriculturalHoldingRadio-2"
                            name="otherTenantsOfAgriculturalHoldingRadio" type="radio" value="no" checked>
                            <label class="govuk-label govuk-radios__label" for="otherTenantsOfAgriculturalHoldingRadio-2">No</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`agricultural-holding GET /other-tenants/change should render the change other tenants of agricultural holding page with "yes" radio option checked if agriculturalHolding.hasOtherTenants is true 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change other tenants of agricultural holding</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="otherTenantsOfAgriculturalHoldingRadio"
                            name="otherTenantsOfAgriculturalHoldingRadio" type="radio" value="yes"
                            checked>
                            <label class="govuk-label govuk-radios__label" for="otherTenantsOfAgriculturalHoldingRadio">Yes</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="otherTenantsOfAgriculturalHoldingRadio-2"
                            name="otherTenantsOfAgriculturalHoldingRadio" type="radio" value="no">
                            <label class="govuk-label govuk-radios__label" for="otherTenantsOfAgriculturalHoldingRadio-2">No</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`agricultural-holding GET /tenant/change should render the change tenant of agricultural holding page with "no" radio option checked if agriculturalHolding.isTenant is false 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change tenant of agricultural holding</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="tenantOfAgriculturalHoldingRadio"
                            name="tenantOfAgriculturalHoldingRadio" type="radio" value="yes">
                            <label class="govuk-label govuk-radios__label" for="tenantOfAgriculturalHoldingRadio">Yes</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="tenantOfAgriculturalHoldingRadio-2"
                            name="tenantOfAgriculturalHoldingRadio" type="radio" value="no" checked>
                            <label class="govuk-label govuk-radios__label" for="tenantOfAgriculturalHoldingRadio-2">No</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;

exports[`agricultural-holding GET /tenant/change should render the change tenant of agricultural holding page with "yes" radio option checked if agriculturalHolding.isTenant is true 1`] = `
"<main class="govuk-main-wrapper " id="main-content" role="main">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full"><span class="govuk-caption-l govuk-!-margin-bottom-3">Appeal 351062</span>
                    <h1                     class="govuk-heading-l govuk-!-margin-bottom-3">Change tenant of agricultural holding</h1>
                </div>
            </div>
            <form method="POST" novalidate="novalidate">
                <div class="govuk-form-group">
                    <div class="govuk-radios" data-module="govuk-radios">
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="tenantOfAgriculturalHoldingRadio"
                            name="tenantOfAgriculturalHoldingRadio" type="radio" value="yes" checked>
                            <label class="govuk-label govuk-radios__label" for="tenantOfAgriculturalHoldingRadio">Yes</label>
                        </div>
                        <div class="govuk-radios__item">
                            <input class="govuk-radios__input" id="tenantOfAgriculturalHoldingRadio-2"
                            name="tenantOfAgriculturalHoldingRadio" type="radio" value="no">
                            <label class="govuk-label govuk-radios__label" for="tenantOfAgriculturalHoldingRadio-2">No</label>
                        </div>
                    </div>
                </div>
                <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
            </form>
        </div>
    </div>
</main>"
`;
